### ISSUE TITLE
DicordClient.DeleteAllMessagesInChannel only deletes messages made since the bot logged in

### ISSUE THREAD

**Issue by CPSibo** [Luigifan/DiscordSharp#10]
I ran into an interesting issue when my group needed to clear out the default discord text channel. Since it's the default one, it couldn't just be deleted and remade. We actually had to delete however many months worth of messages.

Unfortunately, the default behavior of DicordClient.DeleteAllMessagesInChannel() only deletes messages in the client's MessageLog which only seemed to go as far back as when the bot logged in.

I changed the method to

``` c#
public int DeleteAllMessagesInChannel(DiscordChannel channel)
{
    int count = 0;

    // Grab the most recent 100 messages.
    var messages = this.GetMessageHistory(channel, 100, null, null);

    while (messages != null && messages.Count > 0)
    {
        count += messages.Count;

        foreach (var message in messages)
        {
            // Probably a redundant check but might be good to keep.
            if (message.channel == channel)
            {
                this.SendDeleteRequest(message);
            }
        }

        // Grab the most recent 100 messages.
        messages = this.GetMessageHistory(channel, 100, null, null);
    }

    return count;
}
```

I didn't want to make a pull request since I'm not sure if you even intend to support this functionality. But it might be good to at least offer a new method that can truly prune an entire channel.

The only question is how Discord's API will react to that number of requests. If the API does have a flood limit, you could add a sleep to the loop or something. Otherwise, it might be impetus for the Discord devs to add a "Delete all messages in this channel" function to the app or API.

**Comment by suicvne** [comment]
The method was originally intended to only handle since it was logged in as it pulled messages from its internal log and such.

There's a rate limit and currently the bot doesn't handle rate limits (though, I should get around to that once I find out documentation on it).

I like your change however and I think that I will add it in once I do get documentation on rate limits :) Actually, message deletion might not be rate limited as of yet. I have a couple of other things I need to handle with message deletion too, including private message deletion. :)

**Comment by suicvne** [comment]
I partially tested this method with the max value (100) and it seemed to work, though, I coded it differently using Linq because I like Linq :stuck_out_tongue: .

``` csharp
        /// <summary>
        /// Deletes the specified number of messages in a given channel.
        /// Thank you to Siegen for this idea/method!
        /// </summary>
        /// <param name="channel">The channel to delete messages in.</param>
        /// <param name="count">The amount of messages to delete (max 100)</param>
        /// <returns>The count of messages deleted.</returns>
        public int DeleteMultipleMessagesInChannel(DiscordChannel channel, int count)
        {
            if (count > 100)
                count = 100;

            int __count = 0;

            var messages = GetMessageHistory(channel, count, null, null);

            messages.ForEach(x => 
            {
                if (x.channel.id == channel.id)
                {
                    SendDeleteRequest(x);
                    __count++;
                }
            });

            return __count;
        }
```

I ultimately left it up to the user to handle the resting and whatnot. Though, from what I can see, it did a pretty good job of completely wiping out my test channel which _should've_ had more than 100 messages. :scream: 

25a41fc98fc7ee0e24dbd551510db3f237e6a20c is the commit containing this change.

**Comment by CPSibo** [comment]
Yeah, it seems like after maybe 10 minutes of it running, I'd start getting web socket errors which sounded like the server was just giving up. Adding a 2 second sleep to the outer loop seemed to fix it, though.

**Comment by suicvne** [comment]
More than likely yeah that is what it is. 

a589bcfca61a243e72be0f6c67bc447f1bbb3751 has some experimental rate limiting support (but i fixed bugs with it in b4c4fbf8cc79a8133b73582b1ea363e3f913a255) so you might want to try that. As of right now, it internally waits and then resumes sending the message. Hopefully this doesn't screw with anything.
